cl/aggregation/pool_impl.go

```diff
@@ -0,0 +1,229 @@
+	netConfig *clparams.NetworkConfig,
+) AggregationPool {
+	p := &aggregationPoolImpl{
+		genesisConfig:  genesisConfig,
+		beaconConfig:   beaconConfig,
+		netConfig:      netConfig,
+		aggregatesLock: sync.RWMutex{},
+		aggregates:     make(map[[32]byte][]*attestation),
+	}
+	go p.sweepStaleAtt(ctx)
+	return p
+}
+
+func (p *aggregationPoolImpl) AddAttestation(inAtt *solid.Attestation) error {
+	// use hash of attestation data as key
+	hashRoot, err := inAtt.AttestantionData().HashSSZ()
+	if err != nil {
+		return err
+	}
+
+	p.aggregatesLock.Lock()
+	defer p.aggregatesLock.Unlock()
+	if _, ok := p.aggregates[hashRoot]; !ok {
+		p.aggregates[hashRoot] = []*attestation{
+			{
+				bitCount: countBit(inAtt),
+				att:      inAtt,
+			},
+		}
+		return nil
+	}
+
+	// NOTE: naive merge attestation for each existing attestation in the pool,
+	// but it's not optimal. it's kind of a maximum coverage problem.
+	mergeCount := 0
+	alreadyContain := false
+	after := []*attestation{}
+	for _, curAtt := range p.aggregates[hashRoot] {
+		if isSubset(curAtt.att.AggregationBits(), inAtt.AggregationBits()) {
+			// in this case, the new attestation is already contained in the existing attestation, so do not need
+			// to add it again no matter it's merged or not.
+			alreadyContain = true
+			after = append(after, curAtt)
+			continue
+		}
+
+		if overlap, err := checkOverlap(curAtt.att.AggregationBits(), inAtt.AggregationBits()); err != nil {
+			return err
+		} else if overlap {
+			// do nothing but just append the original attestation
+			after = append(after, curAtt)
+		} else {
+			// merge attestation
+			mergedAtt, err := mergeAttestationNoOverlap(inAtt, curAtt.att)
+			if err != nil {
+				return err
+			}
+			after = append(after, &attestation{
+				bitCount: countBit(mergedAtt),
+				att:      mergedAtt,
+			})
+			mergeCount++
+		}
+	}
+	if mergeCount == 0 && !alreadyContain {
+		// no merge and no contain, add new attestation
+		after = append(after, &attestation{
+			bitCount: countBit(inAtt),
+			att:      inAtt,
+		})
+	}
+	p.aggregates[hashRoot] = after
+	return nil
+}
+
+func mergeAttestationNoOverlap(a, b *solid.Attestation) (*solid.Attestation, error) {
+	// merge bit
+	newBits := make([]byte, len(a.AggregationBits()))
+	for i := range a.AggregationBits() {
+		newBits[i] = a.AggregationBits()[i] | b.AggregationBits()[i]
+	}
+	// merge sig
+	aSig := a.Signature()
+	bSig := b.Signature()
+	sig1 := make([]byte, len(aSig))
+	sig2 := make([]byte, len(bSig))
+	copy(sig1, aSig[:])
+	copy(sig2, bSig[:])
+	mergedSig, err := blsAggregate([][]byte{sig1, sig2})
+	if err != nil {
+		return nil, err
+	}
+	if len(mergedSig) > 96 {
+		return nil, fmt.Errorf("merged signature is too long")
+	}
+	var mergedResult [96]byte
+	copy(mergedResult[:], mergedSig)
+	merge := solid.NewAttestionFromParameters(
+		newBits,
+		a.AttestantionData(),
+		mergedResult,
+	)
+	return merge, nil
+}
+
+func isSubset(a, b []byte) bool {
```

Comment:
Maybe this is a bit simpler:

```
func IsSupersetBitlist(a, b []byte) bool {
	if len(a) < len(b) {
		return false
	}
	for i := range b {
		if a[i]&b[i] != b[i] {
			return false
		}
	}
	return true
}
```

this could be useful to be moved to cl/utils package in `bytes.go` perhaps because it will be reused

Reply:
alrdy copied your code.